<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<HTML>
 <HEAD>
   <TITLE> Handshake Failure : SSL_accept:Error in before SSL initialization
   </TITLE>
   <LINK REL="Index" HREF="https://mta.openssl.org/pipermail/openssl-users/2022-February/index.html" >
   <LINK REL="made" HREF="mailto:openssl-users%40openssl.org?Subject=Re%3A%20Handshake%20Failure%20%3A%20SSL_accept%3AError%20in%20before%20SSL%20initialization&In-Reply-To=%3CCAMiDcbHMcEnYvZ95TAbc%2B5Q3AQCVzYUAEgOu2pcNY80oi6YrKg%40mail.gmail.com%3E">
   <META NAME="robots" CONTENT="index,nofollow">
   <style type="text/css">
       pre {
           white-space: pre-wrap;       /* css-2.1, curent FF, Opera, Safari */
           }
   </style>
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="014853.html">
   <LINK REL="Next"  HREF="014851.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>Handshake Failure : SSL_accept:Error in before SSL initialization</H1>
    <B>Kamala Ayyar</B> 
    <A HREF="mailto:openssl-users%40openssl.org?Subject=Re%3A%20Handshake%20Failure%20%3A%20SSL_accept%3AError%20in%20before%20SSL%20initialization&In-Reply-To=%3CCAMiDcbHMcEnYvZ95TAbc%2B5Q3AQCVzYUAEgOu2pcNY80oi6YrKg%40mail.gmail.com%3E"
       TITLE="Handshake Failure : SSL_accept:Error in before SSL initialization">kamala.ayyar at gmail.com
       </A><BR>
    <I>Thu Feb 10 15:41:59 UTC 2022</I>
    <P><UL>
        <LI>Previous message (by thread): <A HREF="014853.html">Handshake Failure : SSL_accept:Error in before SSL initialization
</A></li>
        <LI>Next message (by thread): <A HREF="014851.html">Not able to perform FIPS self-tests
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#14866">[ date ]</a>
              <a href="thread.html#14866">[ thread ]</a>
              <a href="subject.html#14866">[ subject ]</a>
              <a href="author.html#14866">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>Hello Matt,

I used the same test server application to listen on the port but used the
command line version for the client to connect and it connects without
issue and handshake done and server certificates displayed on
screen. openssl s_client -connect servername:20333
I am not sure what is the issue with the test client application written in
c++  which uses the same library and the same certificates.  I am not doing
any Client Authentication. I get the same error -*Error in SSLv3/TLS write
client hello*
The non SSL socket connects and data is exchanged.

Thanks
Kamala



On Tue, Feb 8, 2022 at 1:17 PM Kamala Ayyar &lt;<A HREF="../../../mailman/listinfo/openssl-users.html">kamala.ayyar at gmail.com</A>&gt; wrote:

&gt;<i> Hello Matt,
</I>&gt;<i>
</I>&gt;<i> The socket descriptor is good and I verified using the socket directly.  I
</I>&gt;<i> do exchange data between client and server successfully before passing it
</I>&gt;<i> the function to convert to a secure socket.
</I>&gt;<i> It fails at the same place as the SSL_accept() with the same error  from
</I>&gt;<i> the call back function
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> *SSL_acceptSSL_CB_LOOP | SSL_accept:before SSL initializationSSL_accept*
</I>&gt;<i> *SSL_accept:Error in before SSL initialization*
</I>&gt;<i> On the client side the error is
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> *SSL_connectSSL_CB_LOOP | SSL_connect:before SSL
</I>&gt;<i> initializationSSL_connectSSL_connect:Error in SSLv3/TLS write client hello*
</I>&gt;<i> I used your example and wrapped the socket with the bio however I was not
</I>&gt;<i> able to do a bio read(). I am getting a return of -1 from the bio_read and
</I>&gt;<i> SSL_get_error() and ERR_prints_error does not print anything
</I>&gt;<i>
</I>&gt;<i> Thanks
</I>&gt;<i> Kamala
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i>
</I>&gt;<i> On Fri, Feb 4, 2022 at 1:20 PM Matt Caswell &lt;<A HREF="../../../mailman/listinfo/openssl-users.html">matt at openssl.org</A>&gt; wrote:
</I>&gt;<i>
</I>&gt;&gt;<i> Are you sure that the socket descriptor in &quot;*this&quot; is good and works?
</I>&gt;&gt;<i>
</I>&gt;&gt;<i> You could test that by wrapping it in a BIO like this:
</I>&gt;&gt;<i>
</I>&gt;&gt;<i>      BIO *bio = BIO_new(BIO_s_socket());
</I>&gt;&gt;<i>
</I>&gt;&gt;<i>      if (bio == NULL)
</I>&gt;&gt;<i>         goto err;
</I>&gt;&gt;<i>      BIO_set_fd(bio, *this, BIO_NOCLOSE);
</I>&gt;&gt;<i>
</I>&gt;&gt;<i> and then attempting to read some data from it using BIO_read(). If the
</I>&gt;&gt;<i> BIO_read call fails then it suggests the socket descriptor is bad.
</I>&gt;&gt;<i>
</I>&gt;&gt;<i> Matt
</I>&gt;&gt;<i>
</I>&gt;&gt;<i>
</I>&gt;&gt;<i>
</I>&gt;&gt;<i> On 04/02/2022 18:06, Kamala Ayyar wrote:
</I>&gt;&gt;<i> &gt; Hello Matt,
</I>&gt;&gt;<i> &gt;
</I>&gt;&gt;<i> &gt; I call the WSAGetLastError() for Windows and that returns 183
</I>&gt;&gt;<i> &gt; (ERROR_ALREADY_EXISTS) //Cannot create a file when that file already
</I>&gt;&gt;<i> exists
</I>&gt;&gt;<i> &gt; The SSL_get_error() gives us SSL_ERROR_SYSCALL
</I>&gt;&gt;<i> &gt; *Server *code is roughly like below
</I>&gt;&gt;<i> &gt; SSL_CTX *m_pCtx;
</I>&gt;&gt;<i> &gt; SSL *m_pSsl;
</I>&gt;&gt;<i> &gt; m_pCtx = SSL_CTX_new(TLS_server_method();
</I>&gt;&gt;<i> &gt; if ((dwRet = LoadCertificates()) != rSUCCESS)
</I>&gt;&gt;<i> &gt; throw dwRet;
</I>&gt;&gt;<i> &gt; if ((m_pSsl = SSL_new(m_pCtx)) != NULL)
</I>&gt;&gt;<i> &gt; {
</I>&gt;&gt;<i> &gt;       if ((iRet = SSL_set_fd(m_pSsl, (*this)())) == 0) /* attach the
</I>&gt;&gt;<i> &gt; socket descriptor */
</I>&gt;&gt;<i> &gt;      {
</I>&gt;&gt;<i> &gt;      sslError = SSL_get_error(m_pSsl, iRet);
</I>&gt;&gt;<i> &gt;      LOGERROR(szLine);
</I>&gt;&gt;<i> &gt;      throw eSSL_ERROR;
</I>&gt;&gt;<i> &gt;      }
</I>&gt;&gt;<i> &gt;     SSL_set_info_callback(m_pSsl, apps_ssl_info_callback);
</I>&gt;&gt;<i> &gt;     ERR_clear_error();
</I>&gt;&gt;<i> &gt;     if ((sslError = SSL_accept(m_pSsl)) &lt; 1)
</I>&gt;&gt;<i> &gt;       {
</I>&gt;&gt;<i> &gt;           sslError = SSL_get_error(m_pSsl, sslError);
</I>&gt;&gt;<i> &gt;           dwRet = handleError(sslError, &quot;SSL_accept failed with error
</I>&gt;&gt;<i> &quot;,
</I>&gt;&gt;<i> &gt; iRet);
</I>&gt;&gt;<i> &gt;           throw dwRet;// eSSL_ERROR;
</I>&gt;&gt;<i> &gt;       }
</I>&gt;&gt;<i> &gt; }
</I>&gt;&gt;<i> &gt;
</I>&gt;&gt;<i> &gt; Client
</I>&gt;&gt;<i> &gt; SSL_CTX *m_pCtx;
</I>&gt;&gt;<i> &gt; SSL *m_pSsl;
</I>&gt;&gt;<i> &gt; m_pCtx = SSL_CTX_new(TLS_client_method();
</I>&gt;&gt;<i> &gt; if ((dwRet = LoadCertificates(TRUE)) != rSUCCESS) //Trust certificates
</I>&gt;&gt;<i> only
</I>&gt;&gt;<i> &gt; throw dwRet;
</I>&gt;&gt;<i> &gt; /* Set for server verification*/
</I>&gt;&gt;<i> &gt; SSL_CTX_set_verify(m_pCtx, SSL_VERIFY_PEER, NULL); //Work in progress
</I>&gt;&gt;<i> &gt; m_pSsl = SSL_new(m_pCtx);
</I>&gt;&gt;<i> &gt; if ((iRet = SSL_set_fd(m_pSsl, (*this)())) == 0) /* attach the socket
</I>&gt;&gt;<i> &gt; descriptor */
</I>&gt;&gt;<i> &gt; {
</I>&gt;&gt;<i> &gt;     ssl_error = SSL_get_error(m_pSsl, iRet);
</I>&gt;&gt;<i> &gt;     LOGERROR(szLine);
</I>&gt;&gt;<i> &gt;     throw eSSL_ERROR;
</I>&gt;&gt;<i> &gt; }
</I>&gt;&gt;<i> &gt; SSL_set_info_callback(m_pSsl, apps_ssl_info_callback);
</I>&gt;&gt;<i> &gt; ERR_clear_error();
</I>&gt;&gt;<i> &gt; if ((iRet = SSL_connect(m_pSsl)) &lt;= 0)   /* perform the connection */
</I>&gt;&gt;<i> &gt; {
</I>&gt;&gt;<i> &gt; ssl_error = SSL_get_error(m_pSsl, iRet);
</I>&gt;&gt;<i> &gt; dwRet = handleError(iRet, &quot;SSL_connect failed with error &quot;, ssl_error);
</I>&gt;&gt;<i> &gt; throw eSSL_ERROR;
</I>&gt;&gt;<i> &gt; }
</I>&gt;&gt;<i> &gt;
</I>&gt;&gt;<i> &gt; ShowCerts();
</I>&gt;&gt;<i> &gt; }
</I>&gt;&gt;<i> &gt;
</I>&gt;&gt;<i> &gt; As mentioned before this code works fine when  called by another
</I>&gt;&gt;<i> &gt; application. So the certificates are all valid. I also tried this on
</I>&gt;&gt;<i> &gt; different machines  but it did not work- I get the same error.
</I>&gt;&gt;<i> &gt; Thanks
</I>&gt;&gt;<i> &gt; Kamala
</I>&gt;&gt;<i> &gt;
</I>&gt;&gt;<i> &gt; On Fri, Feb 4, 2022 at 12:20 PM Matt Caswell &lt;<A HREF="../../../mailman/listinfo/openssl-users.html">matt at openssl.org</A>
</I>&gt;&gt;<i> &gt; &lt;mailto:<A HREF="../../../mailman/listinfo/openssl-users.html">matt at openssl.org</A>&gt;&gt; wrote:
</I>&gt;&gt;<i> &gt;
</I>&gt;&gt;<i> &gt;     Does errno give you anything?
</I>&gt;&gt;<i> &gt;
</I>&gt;&gt;<i> &gt;     How did you create your BIOs for m_pSsl?
</I>&gt;&gt;<i> &gt;
</I>&gt;&gt;<i> &gt;     Matt
</I>&gt;&gt;<i> &gt;
</I>&gt;&gt;<i> &gt;     On 04/02/2022 16:25, Kamala Ayyar wrote:
</I>&gt;&gt;<i> &gt;      &gt; Hello Matt,
</I>&gt;&gt;<i> &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt; The SSL_get_error() returns 5(SSL_ERROR_SYSCALL) It does not
</I>&gt;&gt;<i> print
</I>&gt;&gt;<i> &gt;      &gt; anything for this error, just an empty string.
</I>&gt;&gt;<i> &gt;      &gt; I use the following to print error but nothing is printed
</I>&gt;&gt;<i> &gt;      &gt; if ((retVal = SSL_accept(m_pSsl)) &lt; 1)
</I>&gt;&gt;<i> &gt;      &gt; {
</I>&gt;&gt;<i> &gt;      &gt; sslError = SSL_get_error(m_pSsl, retVal);
</I>&gt;&gt;<i> &gt;      &gt; LOGERROR(getOpenSSLError());
</I>&gt;&gt;<i> &gt;      &gt; throw dwRet;// eSSL_ERROR;
</I>&gt;&gt;<i> &gt;      &gt; }
</I>&gt;&gt;<i> &gt;      &gt; string getOpenSSLError()
</I>&gt;&gt;<i> &gt;      &gt; {
</I>&gt;&gt;<i> &gt;      &gt; BIO *bio = BIO_new(BIO_s_mem());
</I>&gt;&gt;<i> &gt;      &gt; ERR_print_errors(bio);
</I>&gt;&gt;<i> &gt;      &gt; char *buf;
</I>&gt;&gt;<i> &gt;      &gt; size_t len = BIO_get_mem_data(bio, &amp;buf);
</I>&gt;&gt;<i> &gt;      &gt; string ret(buf, len);
</I>&gt;&gt;<i> &gt;      &gt; BIO_free(bio);
</I>&gt;&gt;<i> &gt;      &gt; return ret;
</I>&gt;&gt;<i> &gt;      &gt; }
</I>&gt;&gt;<i> &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt; *Kamala  Ayyar*
</I>&gt;&gt;<i> &gt;      &gt; 502 Claremont Ave.
</I>&gt;&gt;<i> &gt;      &gt; Teaneck NJ 07666-2563
</I>&gt;&gt;<i> &gt;      &gt; Tel: (201)530-0861
</I>&gt;&gt;<i> &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt; On Fri, Feb 4, 2022 at 10:54 AM Matt Caswell &lt;<A HREF="../../../mailman/listinfo/openssl-users.html">matt at openssl.org</A>
</I>&gt;&gt;<i> &gt;     &lt;mailto:<A HREF="../../../mailman/listinfo/openssl-users.html">matt at openssl.org</A>&gt;
</I>&gt;&gt;<i> &gt;      &gt; &lt;mailto:<A HREF="../../../mailman/listinfo/openssl-users.html">matt at openssl.org</A> &lt;mailto:<A HREF="../../../mailman/listinfo/openssl-users.html">matt at openssl.org</A>&gt;&gt;&gt; wrote:
</I>&gt;&gt;<i> &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt;     On 04/02/2022 15:17, Kamala Ayyar wrote:
</I>&gt;&gt;<i> &gt;      &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt;      &gt; Hello,
</I>&gt;&gt;<i> &gt;      &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt;      &gt; We are facing a strange handshake failure issue with a
</I>&gt;&gt;<i> test
</I>&gt;&gt;<i> &gt;      &gt;     server and
</I>&gt;&gt;<i> &gt;      &gt;      &gt; client application using OpenSSL in Windows.  We have
</I>&gt;&gt;<i> &gt;     tried with
</I>&gt;&gt;<i> &gt;      &gt;     both
</I>&gt;&gt;<i> &gt;      &gt;      &gt; 1.1.1g and 3.0.1 versions- same problem. We created a Dll
</I>&gt;&gt;<i> to
</I>&gt;&gt;<i> &gt;      &gt;     handle the
</I>&gt;&gt;<i> &gt;      &gt;      &gt; OpenSSL functions- where the SSL context, SSL object and
</I>&gt;&gt;<i> &gt;      &gt;     certificates
</I>&gt;&gt;<i> &gt;      &gt;      &gt; are handled. The certificates are obtained from the
</I>&gt;&gt;<i> &gt;     Windows store
</I>&gt;&gt;<i> &gt;      &gt;     and
</I>&gt;&gt;<i> &gt;      &gt;      &gt; converted to cert and key using PKCS12_parse()
</I>&gt;&gt;<i> &gt;      &gt;      &gt; The server accepts non secure connection from the client
</I>&gt;&gt;<i> &gt;     and then
</I>&gt;&gt;<i> &gt;      &gt;     passes
</I>&gt;&gt;<i> &gt;      &gt;      &gt; the socket to the Dll that calls the TLS_server_method()
</I>&gt;&gt;<i> and
</I>&gt;&gt;<i> &gt;      &gt;     creates the
</I>&gt;&gt;<i> &gt;      &gt;      &gt; SSL context, SSL object and loads the certificates for
</I>&gt;&gt;<i> use. It
</I>&gt;&gt;<i> &gt;      &gt;     however
</I>&gt;&gt;<i> &gt;      &gt;      &gt; fails at SSL_accept(m_pSsl). We use a call
</I>&gt;&gt;<i> &gt;      &gt;      &gt; back SSL_set_info_callback(m_pSsl, apps_ssl_info_callback)
</I>&gt;&gt;<i> &gt;     that
</I>&gt;&gt;<i> &gt;      &gt;     gave us
</I>&gt;&gt;<i> &gt;      &gt;      &gt; the following error information
</I>&gt;&gt;<i> &gt;      &gt;      &gt; SSL_accept:Error in before SSL initialization
</I>&gt;&gt;<i> &gt;      &gt;      &gt; On the client side the same Dll is called with a client
</I>&gt;&gt;<i> &gt;      &gt;      &gt; method TLS_client_method() and the error displayed
</I>&gt;&gt;<i> &gt;      &gt;     is SSL_connect:Error
</I>&gt;&gt;<i> &gt;      &gt;      &gt; in SSLv3/TLS write client hello
</I>&gt;&gt;<i> &gt;      &gt;      &gt; We have confirmed the certificates are good and valid.
</I>&gt;&gt;<i> &gt;      &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt;      &gt; The same Dll called from a different heavily threaded
</I>&gt;&gt;<i> &gt;     application
</I>&gt;&gt;<i> &gt;      &gt;     with
</I>&gt;&gt;<i> &gt;      &gt;      &gt; over 2000+ clients works well and handshake connections
</I>&gt;&gt;<i> &gt;     established
</I>&gt;&gt;<i> &gt;      &gt;      &gt; without issues on a different port number.
</I>&gt;&gt;<i> &gt;      &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt;      &gt; We have also tried to use OpenSSL methods directly
</I>&gt;&gt;<i> &gt;     without using
</I>&gt;&gt;<i> &gt;      &gt;     the Dll
</I>&gt;&gt;<i> &gt;      &gt;      &gt; but we get the same failure.  This was also used with
</I>&gt;&gt;<i> &gt;     server and
</I>&gt;&gt;<i> &gt;      &gt;     client
</I>&gt;&gt;<i> &gt;      &gt;      &gt; on the same machine as well as different machines with the
</I>&gt;&gt;<i> &gt;     same
</I>&gt;&gt;<i> &gt;      &gt;      &gt; outcome.  The non secure communication works fine between
</I>&gt;&gt;<i> the
</I>&gt;&gt;<i> &gt;      &gt;     server and
</I>&gt;&gt;<i> &gt;      &gt;      &gt; the client
</I>&gt;&gt;<i> &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt;     What does SSL_get_error() report after SSL_accept() fails?
</I>&gt;&gt;<i> &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt;     Also please dump the OpenSSL error stack when it fails, e.g.
</I>&gt;&gt;<i> &gt;     using
</I>&gt;&gt;<i> &gt;      &gt;     something like ERR_print_errors_fp(stdout);
</I>&gt;&gt;<i> &gt;      &gt;
</I>&gt;&gt;<i> &gt;      &gt;     Matt
</I>&gt;&gt;<i> &gt;      &gt;
</I>&gt;&gt;<i> &gt;
</I>&gt;&gt;<i>
</I>&gt;<i>
</I>-------------- next part --------------
An HTML attachment was scrubbed...
URL: &lt;<A HREF="https://mta.openssl.org/pipermail/openssl-users/attachments/20220210/6151b86b/attachment.htm">https://mta.openssl.org/pipermail/openssl-users/attachments/20220210/6151b86b/attachment.htm</A>&gt;
</PRE>

<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message (by thread): <A HREF="014853.html">Handshake Failure : SSL_accept:Error in before SSL initialization
</A></li>
	<LI>Next message (by thread): <A HREF="014851.html">Not able to perform FIPS self-tests
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#14866">[ date ]</a>
              <a href="thread.html#14866">[ thread ]</a>
              <a href="subject.html#14866">[ subject ]</a>
              <a href="author.html#14866">[ author ]</a>
         </LI>
       </UL>

<hr>
<a href="../../../mailman/listinfo/openssl-users.html">More information about the openssl-users
mailing list</a><br>
</body></html>
